### ISSUE TITLE
issue with saved relationship system

### ISSUE THREAD

**Issue by mcolominas** [kirschbaum-development/eloquent-power-joins#26]
From what I understand from your code, the relationships are stored in an array, and it does not admit 2 with the same name even though they have different parents, for example:

Product::joinRelationship('translation')->joinRelationship('features.translation')

The previous code only gets the relationships of 'translation' from the first join, and the second join only gets the relation of 'features', ignoring the 'translation'.

The problem comes in this line of code:
```
     /**
     * Checks if the relationship was already joined.
     */
    public function relationshipAlreadyJoined($relation)
    {
        return isset(PowerJoins::$joinRelationshipCache[spl_object_id($this)][$relation]);
    }

    /**
     * Marks the relationship as already joined.
     */
    public function markRelationshipAsAlreadyJoined($relation)
    {
        PowerJoins::$joinRelationshipCache[spl_object_id($this)][$relation] = true;
    }
```

**It can be fixed modifying the following:**
In scopeJoinNestedRelationship update to:
```
$latestRelation = null;
$relationNameWithParents = null;
foreach ($relations as $index => $relationName) {
    ...
    $relationCallback = null;
    $relationNameWithParents = isset($relationNameWithParents) ? $relationNameWithParents . '.' . $relationName : $relationName;
    ...

    if ($this->relationshipAlreadyJoined($relationNameWithParents)) {
        $latestRelation = $relation;
        continue;
    }
    ...
    $this->markRelationshipAsAlreadyJoined($relationNameWithParents);
}
```

**Comment by luisdalmolin** [comment]
Thanks, @mcolominas I'll take a look at this soon.

**Comment by luisdalmolin** [comment]
@mcolominas Can you elaborate a bit more on what the issue is here? Is there a runtime error that happens because the query is wrong, or is the problem in the selected data?

Pasting your initial email here for more clarity:

> Hi, I am using your package and I like it a lot, but there is a problem that I have found in your system where you store the relations.
> 
> For example:
> Table "countries" has a relationship with table "country_translations"
> Table "states" has a relationship with table "cities_translations"
> Table "cities" has a relationship with table "citines_translations"
> 
> For the translations a package is used where it points to these tables with a relation called "translation".
> 
> The problem is when you try to obtain the cities with the name of the state and the country, for example using the following > code:
> 
> City::joinRelationship('state.translation')->joinRelationship('state.country.translation')
> 
> From what I understand by checking your code, as there is a relation called "translation", when it finds the other, it detects that it is the same and ignores it, regardless of the parent of the relation.
> 
> I have been able to solve this in a very ugly way, creating a relation within country with another name that points to the relation "translation".
> 
> I hope that in the future, you can fix this little error.

**Comment by mcolominas** [comment]
The problem is the same as the first example:

`dd(State::joinRelationship('translation')->joinRelationship('country.translation')->toSql());`

Output [1]:

> select `states`.* from `states` inner join `state_translations` on `state_translations`.`state_id` = `states`.`id` and `state_translations`.`locale` = ? inner join `countries` on `states`.`country_id` = `countries`.`id`

Only one translation, but when I use my fix error, output [2]:

> select `states`.* from `states` inner join `state_translations` on `state_translations`.`state_id` = `states`.`id` and `state_translations`.`locale` = ? inner join `countries` on `states`.`country_id` = `countries`.`id` inner join `country_translations` on `country_translations`.`country_id` = `countries`.`id` and `country_translations`.`locale` = ?

You see the difference and the problem?
Output 1 is bad, only take 1 translation.
Output 2 is good, using my fix, take 2 translations.

**Comment by luisdalmolin** [comment]
Yep, I see the issue. Thanks for the explanation, I'll be working on pushing the fix for this soon.

**Comment by luisdalmolin** [comment]
@mcolominas This issue should be fixed in version 2.2.0: https://github.com/kirschbaum-development/eloquent-power-joins/releases/tag/2.2.0. I'm closing this issue, but feel free to open if the problem persists.
